bgpd/bgp_route.c

```diff
@@ -10936,16 +10982,76 @@ DEFPY (show_ip_bgp_json,
 		if (argv_find(argv, argc, "exact-match", &idx))
 			exact_match = 1;
 
+		if (!community)
+			sh_type = bgp_show_type_community_all;
+	}
+
+	if (!all) {
+		/* show bgp: AFI_IP6, show ip bgp: AFI_IP */
 		if (community)
 			return bgp_show_community(vty, bgp, community,
-						  exact_match, afi, safi, uj);
+						  exact_match, afi, safi,
+						  show_flags);
 		else
-			return (bgp_show(vty, bgp, afi, safi,
-					 bgp_show_type_community_all, NULL, uj,
-					 wide));
-	}
+			return (bgp_show(vty, bgp, afi, safi, sh_type, NULL,
+					 show_flags));
+	} else {
+		/* show <ip> bgp ipv4 all: AFI_IP, show <ip> bgp ipv6 all:
+		 * AFI_IP6 */
+		if (CHECK_FLAG(show_flags, BGP_SHOW_OPT_AFI_IP)
+		    || CHECK_FLAG(show_flags, BGP_SHOW_OPT_AFI_IP6)) {
+			afi = CHECK_FLAG(show_flags, BGP_SHOW_OPT_AFI_IP)
+				      ? AFI_IP
+				      : AFI_IP6;
+			FOREACH_SAFI(safi) {
+				if (strmatch(get_afi_safi_str(afi, safi, true),
+					     "Unknown"))
+					continue;
+
+				if (!bgp_afi_safi_peer_exists(bgp, afi, safi))
+					continue;
+
+				if (!uj)
+					vty_out(vty,
+						"\nFor address family: %s\n\n",
+						get_afi_safi_str(afi, safi,
+								 false));
+
+				if (community)
+					bgp_show_community(vty, bgp, community,
+							   exact_match, afi,
+							   safi, show_flags);
+				else
+					bgp_show(vty, bgp, afi, safi, sh_type,
+						 NULL, show_flags);
+			}
+		} else {
+			/* show <ip> bgp all: for each AFI and SAFI*/
+			FOREACH_AFI_SAFI (afi, safi) {
+				if (strmatch(get_afi_safi_str(afi, safi, true),
+					     "Unknown"))
+					continue;
 
-	return bgp_show(vty, bgp, afi, safi, sh_type, NULL, uj, wide);
+				if (!bgp_afi_safi_peer_exists(bgp, afi, safi))
+					continue;
+
+				if (!uj)
+					vty_out(vty,
+						"\nFor address family: %s\n\n",
+						get_afi_safi_str(afi, safi,
+								 false));
+
+				if (community)
+					bgp_show_community(vty, bgp, community,
+							   exact_match, afi,
+							   safi, show_flags);
+				else
+					bgp_show(vty, bgp, afi, safi, sh_type,
```

Comment:
I'm pretty sure this will produce invalid JSON because every time it would be printed as a separate JSONs for each afi/safi without a proper concat.

Reply:
addressed this comment. now it is producing valid json. Please check the sample JSON format once and confirm if its ok.
{
 "vrfId": 0,
 "vrfName": "default",
 "tableVersion": 6,
 "routerId": "1.1.1.1",
 "defaultLocPrf": 100,
 "localAS": 1,
 "routes": { "1.1.1.1/32": [
  {
    "valid":true,
    "bestpath":true,
    "pathFrom":"external",
    "prefix":"1.1.1.1",
    "prefixLen":32,
    "network":"1.1.1.1\/32",
    "metric":0,
    "weight":32768,
    "peerId":"(unspec)",
    "path":"",
    "origin":"incomplete",
    "nexthops":[
      {
        "ip":"0.0.0.0",
        "hostname":"router",
        "afi":"ipv4",
        "used":true
      }
    ]
  }
],"2.2.2.2/32": [
  {
    "valid":true,
    "bestpath":true,
    "pathFrom":"internal",
    "prefix":"2.2.2.2",
    "prefixLen":32,
    "network":"2.2.2.2\/32",
    "metric":0,
    "locPrf":100,
    "weight":0,
    "peerId":"192.168.56.152",
    "path":"",
    "origin":"incomplete",
    "nexthops":[
      {
        "ip":"192.168.56.152",
        "hostname":"router",
        "afi":"ipv4",
        "used":true
      }
    ]
  }
],"10.0.2.0/24": [
  {
    "valid":true,
    "pathFrom":"internal",
    "prefix":"10.0.2.0",
    "prefixLen":24,
    "network":"10.0.2.0\/24",
    "metric":0,
    "locPrf":100,
    "weight":0,
    "peerId":"192.168.56.152",
    "path":"",
    "origin":"incomplete",
    "nexthops":[
      {
        "ip":"192.168.56.152",
        "hostname":"router",
        "afi":"ipv4",
        "used":true
      }
    ]
  },
  {
    "valid":true,
    "bestpath":true,
    "pathFrom":"external",
    "prefix":"10.0.2.0",
    "prefixLen":24,
    "network":"10.0.2.0\/24",
    "metric":0,
    "weight":32768,
    "peerId":"(unspec)",
    "path":"",
    "origin":"incomplete",
    "nexthops":[
      {
        "ip":"0.0.0.0",
        "hostname":"router",
        "afi":"ipv4",
        "used":true
      }
    ]
  }
],"192.168.56.0/24": [
  {
    "valid":true,
    "pathFrom":"internal",
    "prefix":"192.168.56.0",
    "prefixLen":24,
    "network":"192.168.56.0\/24",
    "metric":0,
    "locPrf":100,
    "weight":0,
    "peerId":"192.168.56.152",
    "path":"",
    "origin":"incomplete",
    "nexthops":[
      {
        "ip":"192.168.56.152",
        "hostname":"router",
        "afi":"ipv4",
        "used":true
      }
    ]
  },
  {
    "valid":true,
    "bestpath":true,
    "pathFrom":"external",
    "prefix":"192.168.56.0",
    "prefixLen":24,
    "network":"192.168.56.0\/24",
    "metric":0,
    "weight":32768,
    "peerId":"(unspec)",
    "path":"",
    "origin":"incomplete",
    "nexthops":[
      {
        "ip":"0.0.0.0",
        "hostname":"router",
        "afi":"ipv4",
        "used":true
      }
    ]
  }
],"192.168.123.245/32": [
  {
    "valid":true,
    "bestpath":true,
    "pathFrom":"internal",
    "prefix":"192.168.123.245",
    "prefixLen":32,
    "network":"192.168.123.245\/32",
    "metric":0,
    "locPrf":100,
    "weight":0,
    "peerId":"192.168.56.152",
    "path":"",
    "origin":"incomplete",
    "nexthops":[
      {
        "ip":"192.168.56.152",
        "hostname":"router",
        "afi":"ipv4",
        "used":true
      }
    ]
  }
],"192.168.223.245/32": [
  {
    "valid":true,
    "bestpath":true,
    "pathFrom":"internal",
    "prefix":"192.168.223.245",
    "prefixLen":32,
    "network":"192.168.223.245\/32",
    "metric":0,
    "locPrf":100,
    "weight":0,
    "peerId":"192.168.56.152",
    "path":"",
    "origin":"incomplete",
    "nexthops":[
      {
        "ip":"192.168.56.152",
        "hostname":"router",
        "afi":"ipv4",
        "used":true
      }
    ]
  }
] } ,
 "vrfId": 0,
 "vrfName": "default",
 "tableVersion": 0,
 "routerId": "1.1.1.1",
 "defaultLocPrf": 100,
 "localAS": 1,
 "routes": {  } ,
 "vrfId": 0,
 "vrfName": "default",
 "tableVersion": 3,
 "routerId": "1.1.1.1",
 "defaultLocPrf": 100,
 "localAS": 1,
 "routes": { "2001:db8::1/128": [
  {
    "valid":true,
    "bestpath":true,
    "pathFrom":"external",
    "prefix":"2001:db8::1",
    "prefixLen":128,
    "network":"2001:db8::1\/128",
    "metric":0,
    "weight":32768,
    "peerId":"(unspec)",
    "path":"",
    "origin":"incomplete",
    "nexthops":[
      {
        "ip":"::",
        "hostname":"router",
        "afi":"ipv6",
        "scope":"global",
        "used":true
      }
    ]
  }
],"2001:db8::2/128": [
  {
    "valid":true,
    "bestpath":true,
    "pathFrom":"internal",
    "prefix":"2001:db8::2",
    "prefixLen":128,
    "network":"2001:db8::2\/128",
    "metric":0,
    "locPrf":100,
    "weight":0,
    "peerId":"192.168.56.152",
    "path":"",
    "origin":"incomplete",
    "nexthops":[
      {
        "ip":"fe80::a00:27ff:fefc:2aa",
        "hostname":"router",
        "afi":"ipv6",
        "scope":"global",
        "used":true
      },
      {
        "ip":"fe80::a00:27ff:fefc:2aa",
        "hostname":"router",
        "afi":"ipv6",
        "scope":"link-local"
      }
    ]
  }
],"2001:db8:85a3::8a2e:370:7334/128": [
  {
    "valid":true,
    "bestpath":true,
    "pathFrom":"external",
    "prefix":"2001:db8:85a3::8a2e:370:7334",
    "prefixLen":128,
    "network":"2001:db8:85a3::8a2e:370:7334\/128",
    "metric":0,
    "weight":32768,
    "peerId":"(unspec)",
    "path":"",
    "origin":"incomplete",
    "nexthops":[
      {
        "ip":"::",
        "hostname":"router",
        "afi":"ipv6",
        "scope":"global",
        "used":true
      }
    ]
  }
] } ,
 "vrfId": 0,
 "vrfName": "default",
 "tableVersion": 0,
 "routerId": "1.1.1.1",
 "defaultLocPrf": 100,
 "localAS": 1,
 "routes": {  } }

Reply:
addressed this comment and added the sample JSON output in commit message.